File: week8/p9.py
class Solution:
    def findMin(self, root):
        while root.left:
            root = root.left
        return root

    def deleteNode(self, root, x):
        if not root:
            return root
        if x < root.data:
            root.left = self.deleteNode(root.left, x)
        elif x > root.data:
            root.right = self.deleteNode(root.right, x)
        else:
            if not root.left:
                return root.right
            elif not root.right:
                return root.left
            temp = self.findMin(root.right)
            root.data = temp.data
            root.right = self.deleteNode(root.right, temp.data)
        return root

File: week8/p9.c
/*
struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};
*/
struct Node* findMin(struct Node* root) {
    while (root && root->left != NULL) {
        root = root->left;
    }
    return root;
}

struct Node* deleteNode(struct Node* root, int x) {
    if (root == NULL) return root;

    if (x < root->data) {
        root->left = deleteNode(root->left, x);
    } else if (x > root->data) {
        root->right = deleteNode(root->right, x);
    } else {
        if (root->left == NULL) {
            struct Node* temp = root->right;
            free(root);
            return temp;
        } else if (root->right == NULL) {
            struct Node* temp = root->left;
            free(root);
            return temp;
        }
        struct Node* temp = findMin(root->right);
        root->data = temp->data;
        root->right = deleteNode(root->right, temp->data);
    }
    return root;
}
